File: Realizable.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package realizable;

import java.util.ArrayList;
import java.util.Scanner;
import java.util.Stack;



class Node
{
    int val;
    Node left;
    Node right;
    Node prev;
}



public class Realizable {

    static ArrayList<Node> leaves = new ArrayList();
    
    public static boolean realizable(int[] A, int T)
    {
        int S = 0;
        for(int i = 1; i < A.length; i++)
            S += A[i];
        //System.out.println(S);
        boolean[][] P = new boolean[A.length][2*S+1];
        P[0][S] = true; 
        
        for(int i = 1; i < P.length; i++)
        {
            for(int j = -S; j <= S; j++)
            {
                if((j - A[i] <= S) && (j - A[i] >= (-1*S)) && P[i-1][j - A[i] + S])
                    P[i][j + S] = true;
                if((j + A[i] <= S) && (j + A[i] >= (-1*S)) && P[i-1][j + A[i] + S])
                    P[i][j+S] = true;
            }
        }
            
        return P[A.length-1][T+S];
    }
    
    public static void showone(int[] A, int T)
    {
        int S = 0;
        for(int i = 1; i < A.length; i++)
            S += A[i];
        //System.out.println(S);
        boolean[][] P = new boolean[A.length][2*S+1];
        P[0][S] = true; 
        
        for(int i = 1; i < P.length; i++)
        {
            for(int j = -S; j <= S; j++)
            {
                if((j - A[i] <= S) && (j - A[i] >= (-1*S)) && P[i-1][j - A[i] + S])
                    P[i][j + S] = true;
                if((j + A[i] <= S) && (j + A[i] >= (-1*S)) && P[i-1][j + A[i] + S])
                    P[i][j+S] = true;
            }
        }
        
        
        if(P[A.length-1][T+S])
        {
            Stack s = new Stack();
            int j = T;
            for(int i = A.length-1; i > 0; i--)
            {
                if((j - A[i] <= S) && (j - A[i] >= (-1*S)) && P[i-1][j - A[i] + S])
                {
                    s.push(new String("-" + A[i]));
                    j += A[i];
                }
                else
                {
                    s.push(new String("+" + A[i]));
                    j -= A[i];
                }
            }
            String str = "Solution: ";
            while (!s.isEmpty())
                str += s.pop();
            System.out.println(str += " = " + T);
        }
        else
            System.out.printf("The value %d is not realizable",T);
        
    }
    
    
    public static void showall(int[] A, int T)
    {
        int S = 0;
        for(int i = 1; i < A.length; i++)
            S += A[i];
        //System.out.println(S);
        boolean[][] P = new boolean[A.length][2*S+1];
        P[0][S] = true; 
        
        for(int i = 1; i < P.length; i++)
        {
            for(int j = -S; j <= S; j++)
            {
                if((j - A[i] <= S) && (j - A[i] >= (-1*S)) && P[i-1][j - A[i] + S])
                    P[i][j + S] = true;
                if((j + A[i] <= S) && (j + A[i] >= (-1*S)) && P[i-1][j + A[i] + S])
                    P[i][j+S] = true;
            }
        }
        
        
        if(P[A.length-1][T+S])
        {
            Node root = new Node();
            root.val = T;
            buildTree(A,A.length-1,P,T,S,root);
            printSolutions(T);
        }
        else
            System.out.println("Number of solutions = 0");
        
    }
    
    public static void buildTree(int[] A,int i,boolean[][] P, int j, int S, Node root)
    {
        //System.out.println(root.val + " " + i);
        if(i == 0)
        {
            leaves.add(root);
            return;
        }
        if((j + A[i] <= S) && (j + A[i] >= (-1*S)) && P[i-1][j + A[i] + S])
                {
                    Node n = new Node();
                    root.right = n;
                    //j += A[i];
                    n.val = j + A[i];
                    n.prev = root;
                    buildTree(A,i-1,P,n.val,S,n);
                }
        if((j - A[i] <= S) && (j - A[i] >= (-1*S)) && P[i-1][j - A[i] + S])
                {
                    Node n = new Node();
                    root.left = n;
                    //j -= A[i];
                    n.val = j - A[i];
                    n.prev = root;
                    buildTree(A,i-1,P,n.val,S,n);
                }
    
    }
    
    public static void printSolutions(int T)
    {
        System.out.printf("Number of solutions = %d\n",leaves.size());
        
        for(int i = 0; i < leaves.size(); i++)
        {
            Node n = leaves.get(i);
            String str = "";
            while(n.prev != null)
            {
                str += n.prev.val > n.val?"+" + (n.prev.val-n.val):(n.prev.val-n.val);
                n = n.prev;
            }
            System.out.printf("Sol   %d : %s = %d\n",i+1,str,T);
        }
    }
    
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        System.out.println("Enter n: ");
        int n = scan.nextInt();
        int[]  A = new int[n+1];
        //int sum = 0;
        System.out.println("Input array: ");
        for(int i = 1; i < n+1; i++)
        {
            A[i] = scan.nextInt();
            //sum += A[i];
        }
        //System.out.println(sum);
        System.out.println("Enter T: ");
        int T = scan.nextInt();
        System.out.println("n = " + n);
        System.out.println("The input array:");
        for(int i = 1; i < n+1; i++)
        {
            System.out.print(A[i] + " ");
        }
        System.out.println("");
        System.out.println("T = " + T);
        System.out.println("");
        
        System.out.println("Part 1: Realizability check");
        if(realizable(A,T))
            System.out.printf("The value %d is realizable\n\n",T);
        else
            System.out.printf("The value %d is not realizable\n\n",T);
       
        System.out.println("Part 2: One solution");
        showone(A,T);
        
        System.out.println("\n\n");
        System.out.println("Part 3: All solutions");
        showall(A,T);
    }
    
}

File: realizable.py
# -*- coding: utf-8 -*-
"""
Created on Mon Jun  3 16:12:14 2019

@author: Uzair
"""

class Node:

    def __init__(self, val):

        self.left = None
        self.right = None
        self.prev = None
        self.val = val



leaves = []
def realizable(A,T):
    S = 0
    for i in A:
        S += i
    rows,cols = (len(A),2*S+1)
    #print(S)
    P = [[False for i in range(cols)] for j in range(rows)]
    P[0][S] = True
    #print(P)
    for i in range(1,len(P)):
        for j in range(-S,S+1):
            if j-A[i] <= S and j-A[i] >= -S and P[i-1][j-A[i] + S]:
                #print(i,j)
                P[i][j + S] = True
            elif j+A[i] <= S and j+A[i] >= -S and P[i-1][j+A[i] + S]:
                #print(i,j)
                P[i][j + S] = True
    return P[len(A)-1][T+S]


def showone(A,T):
    S = 0
    for i in A:
        S += i
    rows,cols = (len(A),2*S+1)
    #print(S)
    P = [[False for i in range(cols)] for j in range(rows)]
    P[0][S] = True
    #print(P)
    for i in range(1,len(P)):
        for j in range(-S,S+1):
            if j-A[i] <= S and j-A[i] >= -S and P[i-1][j-A[i] + S]:
                #print(i,j)
                P[i][j + S] = True
            elif j+A[i] <= S and j+A[i] >= -S and P[i-1][j+A[i] + S]:
                #print(i,j)
                P[i][j + S] = True
    if P[len(A)-1][T+S]:
        j = T
        stack = []
        for i in range(len(A)-1,0,-1):
            if j-A[i] <= S and j-A[i] >= -S and P[i-1][j-A[i] + S]:
                stack.append('-'+str(A[i]))
                j += A[i]
            else:
                stack.append('+'+str(A[i]))
                j -= A[i]
        stri = "Solution: "
        while len(stack) > 0:
            stri += stack.pop()
        print(stri + " = " + str(T))
    else:
        print("No solution")
        


def showall(A,T):
    S = 0
    for i in A:
        S += i
    rows,cols = (len(A),2*S+1)
    #print(S)
    P = [[False for i in range(cols)] for j in range(rows)]
    P[0][S] = True
    #print(P)
    for i in range(1,len(P)):
        for j in range(-S,S+1):
            if j-A[i] <= S and j-A[i] >= -S and P[i-1][j-A[i] + S]:
                #print(i,j)
                P[i][j + S] = True
            elif j+A[i] <= S and j+A[i] >= -S and P[i-1][j+A[i] + S]:
                #print(i,j)
                P[i][j + S] = True
    if P[len(A)-1][T+S]:
        root = Node(T)
        buildTree(A,len(A)-1,P,T,S,root)
        printSolutions(T)
    else:
        print("Number of solutions = 0")
    


def buildTree(A,i,P,j,S,root):
    if i == 0:
        leaves.append(root)
        return
    if j+A[i] <= S and j+A[i] >= -S and P[i-1][j+A[i] + S]:
        n = Node(j+A[i])
        root.right = n
        n.prev = root
        buildTree(A,i-1,P,n.val,S,n)
    if j-A[i] <= S and j-A[i] >= -S and P[i-1][j-A[i] + S]:
        n = Node(j-A[i])
        root.left = n
        n.prev = root
        buildTree(A,i-1,P,n.val,S,n)


def printSolutions(T):
    print("Number of solutions: " + str(len(leaves)))
    j = 0
    for i in leaves:
        j+=1
        stri = ""
        n = i
        while n.prev != None:
            stri += "+" + str(n.prev.val-n.val) if n.prev.val > n.val else str(n.prev.val-n.val)
            n = n.prev
        print("Sol " + str(j) + " " + stri + " = " + str(T))

def main():
    n = int(input("n:"))
    A = list(map(int,input("Space separated numbers: ").split()))
    A = [0] + A
    T = int(input("T: "))
    #print(A)
    print("n = " + str(n))
    print()
    print("Part 1: Realizability check")
    if realizable(A,T):
        print("The value " + str(T) + " is realizable")
    else:
        print("The value " + str(T) + " is not realizable")
    print()
    print("Part 2: One solution")
    showone(A,T)
    print()
    print("Part 3: All solutions")
    showall(A,T)
    
main()

    